File: day2/day2.py
#!/usr/bin/env python3

def parse_data(txt):
    rows = []
    for line in txt.splitlines():
        rows.append(list(map(int, line.split("\t"))))
    return rows

def checksum1(rows):
    value = 0
    for row in rows:
        value += max(row) - min(row)
    return value

def checksum2(rows):
    value = 0
    for row in rows:
        for i in range(1,len(row)):
            for j in range(i):
                if row[i] % row[j] == 0:
                    value += row[i] // row[j]
                elif row[j] % row[i] == 0:
                    value += row[j] // row[i]
    return value

# assert checksum1(parse_data("""5\t1\t9\t5
# 7\t5\t3
# 2\t4\t6\t8""")) == 18

# assert checksum2(parse_data("""5\t9\t2\t8
# 9\t4\t7\t3
# 3\t8\t6\t5""")) == 9

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    rows = parse_data(txt)
    print("Part1:", checksum1(rows))
    print("Part2:", checksum2(rows))

File: day2/day2.cpp
#include <algorithm>
#include <numeric>
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static size_t checksum1(const vector<vector<int>>& rows)
{
	return accumulate(
		rows.begin(),
		rows.end(),
		0ULL,
		[](auto acc, const auto& row) {
			auto [min, max] = minmax_element(row.begin(), row.end());
			if (min != row.end())
			{
				return acc + *max - *min;
			}
			else
			{
				return acc;
			}
		});
}

static size_t checksum2(const vector<vector<int>>& rows)
{
	return accumulate(
		rows.begin(),
		rows.end(),
		0ULL,
		[](auto acc, const auto& row) {
			for (size_t i = 1; i < row.size(); i++)
			{
				for (size_t j = 0; j < i; j++)
				{
					if (row[i] % row[j] == 0)
					{
						return acc + row[i] / row[j];
					}
					else if (row[j] % row[i] == 0)
					{
						return acc + row[j] / row[i];
					}
				}
			}
			return acc;
		});
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}

	vector<vector<int>> rows;
	string line;
	while (getline(input, line))
	{
		istringstream sin(line);
		vector<int> numbers;
		int number;
		while (sin>>number)
		{
			numbers.push_back(number);
		}
		rows.emplace_back(move(numbers));
	}
	input.close();

	cout << fmt::format("Part1: {}", checksum1(rows)) << endl
	     << fmt::format("Part2: {}", checksum2(rows)) << endl;
	return 0;
}
